enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,989 @@
+      # cluster site and plugs the USB stick.
+      digest: sha256:....
+
+status:
+  desired:
+    bundle:
+      # Location of the upgrade bundle. This is populated by the monitor, if
+      # enabled, or copied from the `spec.desiredUpdate.bundle.file` field if
+      # monitoring is disabled.
+      file: /root/my-bunde.tar # Can be a file.
+      file: /dev/sdb # Or a device.
+
+      # Node where the bundle is available. This is optional. By default the
+      # cluster version operator will look for the bundle in all the nodes, but
+      # if this is used it will look only in the given node.
+      node: node2
+
+      # Contains the current status of the bundle processing in each node
+      # of the cluster. Keys are the node names.
+      nodes:
+        node0:
+          # Indicates if the upgrade bundle has already been extracted to
+          # the `/var/lib/upgrade/4.13.7-x86_64` directory.
+          extracted: true|false
+          # Indicates if the images from the upgrade bundle have been already
+          # copied to the `/etc/containers/storage` directory.
+          loaded: true|false
+          # Contains the metadata of the bundle as seen by this node. This is
+          # taken from the `metadata.json` file that is part of the bundle.
+          metadata: '{ ... }'
+        node1: { ... }
+        ...
+```
+
+And this is a summary of the new `ContainerRuntimeConfig` fields:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+  name: pin-upgrade
+spec:
+  containerRuntimeConfig:
+
+    # Additional image store directories. Translates into something like this
+    # inside the `/etc/containers/storage.conf` file:
+    #
+    # additionalimagestores = [
+    #   "/var/lib/my-images/...",
+    #   "/var/lib/your-images/...",
+    #   ...
+    # ]
+    additionalImageStores:
+    - /var/lib/my-images
+    - /var/lib/my-images
+
+    # List of pinned and pre-loaded images. Translates into a new `pin-upgrade.conf`
+    # file inside the `/etc/crio/crio.conf.d` directory with content similar to this:
+    #
+    # pinned_images = [
+    #   "quay.io/openshift-release-dev/ocp-release@sha256:...",
+    #   "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+    #   "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+    #   ...
+    # ]
+    #
+    # The CRI-O service will be reloaded when this is changed.
+    #
+    # The machine config operator will ensure that these images are pre-loadd and
+    # in all the nodes of the cluster.
+    pinnedImages:
+    - quay.io/openshift-release-dev/ocp-release@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    ...
+
+    # Optional directory containing the images. When this is used the machine config
+    # operator will expect it to contain a dump of a registry server, will start a
```

Comment:
Hmmm...a lot of duplication there.  I think we should be able to synthesize the necessary backing data from the image store, and serve it as a registry, no?

Reply:
It guess possible to take an image from a `containers-storage` directory and serve it using the registry protocol. It wouldn't be trivial, but not impossible either. The fact that `skopeo copy containers-storage:...` works means that there is enough information in the containers storage directory to reconstruct what the registry protocol requires. Is that what you are suggesting?

But even if that works it would mean that we bundle would need to use the `container-storage` format, which requires twice the space of a registry dump.

Reply:
It looks like the skopeo `dir` transport may be what we want here.

Reply:
I considered that, but was hesitant to use it because of this statement in the `containers-transport` manual page:

> An existing local directory path storing the manifest, layer tarballs and signatures as individual files.  This is a non-standardized format, primarily useful  for  debugging  or noninvasive container inspection.

Reply:
It's not standardized, but I believe it's stable, i.e. it will not change in an incompatible way.

We can try to get that guaranteed upstream.  And also, this should be covered by CI, so if it *does* change it at least will be something we find and can fix (e.g. by vendoring an older code).

Reply:
Lets say that we use that `dir` format. We will still need one of the following alternatives:

1. A registry server implementation that reads from that directory, so that we can use `crictl pull` (actually the equivalent CRI-O API) to pull the images.

2. Copy the images from that directory to the containers storage directory running `skopeo copy dir:... containers-storage:...` for each image.

The only benefit I see in any of those options is that we will not be depending on the  the `docker/distribution` code. Is that what you are concerned about?

Reply:
Right; it seems valuable to me to not hardcode docker/distribution; but I also wouldn't call doing so a blocker. I care more about using the registry protocol on the wire instead of inventing something new.